model/src/application/interaction/application_command/data/mod.rs

```diff
@@ -26,50 +30,168 @@ pub struct CommandData {
+                        "invalid option data: {:?} has value instead of options",
+                        raw.kind
+                    )));
+                }
+                (kind, value) => {
+                    return Err(de::Error::custom(format!(
+                        "invalid option value/type pair: value is {:?} but type is {:?}",
+                        value, kind,
+                    )));
+                }
+            }
+        } else {
+            let options = raw
+                .options
+                .ok_or_else(|| de::Error::missing_field("options"))?;
+            match raw.kind {
+                CommandOptionType::SubCommand => CommandOptionValue::SubCommand(options),
+                CommandOptionType::SubCommandGroup => CommandOptionValue::SubCommandGroup(options),
+                kind => {
+                    return Err(de::Error::custom(format!(
+                        "no `value` but type is {:?}",
+                        kind
+                    )))
+                }
+            }
+        };
+        Ok(CommandDataOption {
+            name: raw.name,
+            value,
+        })
     }
+}
 
-    pub fn name(&self) -> &str {
+impl CommandOptionValue {
+    const fn kind(&self) -> CommandOptionType {
         match self {
-            Self::String { name, .. }
-            | Self::Integer { name, .. }
-            | Self::Boolean { name, .. }
-            | Self::SubCommand { name, .. } => name,
+            CommandOptionValue::String(_) => CommandOptionType::String,
+            CommandOptionValue::Integer(_) => CommandOptionType::Integer,
+            CommandOptionValue::Boolean(_) => CommandOptionType::Boolean,
+            CommandOptionValue::User(_) => CommandOptionType::User,
+            CommandOptionValue::Channel(_) => CommandOptionType::Channel,
+            CommandOptionValue::Role(_) => CommandOptionType::Role,
+            CommandOptionValue::Mentionable(_) => CommandOptionType::Mentionable,
+            CommandOptionValue::SubCommand(_) => CommandOptionType::SubCommand,
+            CommandOptionValue::SubCommandGroup(_) => CommandOptionType::SubCommandGroup,
         }
     }
+
+    fn raw_value(&self) -> Option<CommandOptionValueRaw<'_>> {
+        Some(match *self {
+            CommandOptionValue::String(ref s) => CommandOptionValueRaw::String(s.into()),
+            CommandOptionValue::Integer(i) => CommandOptionValueRaw::Integer(i),
+            CommandOptionValue::Boolean(b) => CommandOptionValueRaw::Boolean(b),
+            CommandOptionValue::User(UserId(id))
+            | CommandOptionValue::Channel(ChannelId(id))
+            | CommandOptionValue::Role(RoleId(id))
+            | CommandOptionValue::Mentionable(id) => {
+                CommandOptionValueRaw::String(id.to_string().into())
+            }
+            CommandOptionValue::SubCommand(_) | CommandOptionValue::SubCommandGroup(_) => {
+                return None;
```

Comment:
Maybe, we can just have a variant for subcommands in `CommandOptionValueRaw` and then just get rid of the `Option`

Reply:
I think this isn't necessary after b93a3b2?